File: may/bit.py
from math import floor,log2

def fun(x,y,z):
    return (x&z)*(y&z)

def fun2(v,x,y,l,r):
    ans=fun(x,y,v[0])
    idx=v[0]
    for i in range(1,len(v),1):
        temp=fun(x,y,v[i])
        if temp > ans:
            ans=temp
            idx=v[i]
        elif fun(x,y,v[i])==ans and v[i]<idx:
            ans=temp
            idx=v[i]
    
    return idx
    


for _ in range(int(input())):
    x,y,l,r=map(int,input().split())
    if x==0 or y==0:
        print(l)
        continue

    if l==r:
        print(r)
        continue
        
    ans=x|y

    # print(ans)
    if ans <=r:
        print(ans)
        continue
    
    if r>0:
        max_bit=floor(log2(r))
    else:
        max_bit=0
    
    if ans==0:
        print(l)
        continue
        
    if ans>0:
        ans_bit=floor(log2(ans))
    else:
        ans_bit=0

    i=ans_bit
    while i>max_bit:
        check=(1<<i)
        if (ans & check) != 0:
            ans = ans - check
        
        i=i-1
    
    if ans <=r:
        # print(ans)
        # print("Hello")
        if ans&x==0 or ans&y==0 :
            print(0)
        else:
            print(ans)
        continue
    
    pos=[0]
    i=max_bit
    for i in range(max_bit,-1,-1):
        check=(1<<i)
        if((r&check)!=0  and (ans&check)!=0):
            temp=ans-check
            if temp<=r :
                pos.append(temp)
        elif (r&check)==0  and (ans&check)!=0:
            ans = ans - check
            if ans <= r:
                pos.append(ans)         
    print(fun2(pos,x,y,l,r))
    # print(pos)

File: may/bit.cpp
#include <bits/stdc++.h>
#define ios ios_base::sync_with_stdio(false);cin.tie(NULL)
#define int long long
#define mod 1000000007
#define INF 1000000000
#define endl '\n'
#define cout(x) cout<<x<<" "
#define debug(x) cout<<x<<"\n\n"
#define cin(x) cin>>x
#define cin2(x,y) cin>>x>>y
#define cin3(x,y,z) cin>>x>>y>>z


using namespace std;
int fun(int x,int y,int z){
    return ((x&z)*(y&z));
}
int fun2(vector<int> v,int x,int y,int l,int r){
    int ans=fun(x,y,v[0]);
    int id=v[0];
    for(int i=1;i<v.size();i++){
        int temp=fun(x,y,v[i]);
        if(temp>ans){
            ans=temp;
            id=v[i];
        }else if(temp==ans and v[i]<id){
            ans=temp;
            id=v[i];
        }
    }
    return id;
}

int32_t main(){
    ios;
    int t;
    cin>>t;
    while(t--){
        int x,y,l,r;
        cin>>x>>y>>l>>r;
        if(!x or !y){
            cout<<l<<endl;
            continue;
        }

        int ans=x|y;
        
        if(ans<=r){
            cout<<ans<<endl;
            continue;
        }
        int max_bit=floor(log2(r)),ans_bit=floor(log2(ans));
        for(int i=ans_bit;i>max_bit;i--){
            if((ans & (1<<i))!=0){
                ans-=(1<<i);
            }
        }
        if(ans<=r){
            cout<<ans<<endl;
            continue;
        }

        int final=ans-(1<<max_bit);
        vector<int> pos;
        for(int i=max_bit;i>=0;i--){
            int check=(1<<i);
            if((r&check)!=0  and (ans&check)!=0){
                int temp=ans-check;
                if(temp<=r){
                    pos.push_back(temp);
                }       
            }else if((r&check)==0  and (ans&check)!=0){
                ans-=check;
                if(ans<=r)
                    pos.push_back(ans);
            }
        }
        cout<<fun2(pos,x,y,l,r)<<"\n";
        // for(auto i:pos){
        //     cout<<i<<" ";
        // }
        // cout<<endl;
    }
    return 0;
}